### ISSUE TITLE
Listen for visibilitychange, instead of unload/beforeunload

### ISSUE THREAD

**Issue by igrigorik** [googleanalytics/autotrack#6]
> The sessionDurationTracker plugin partially solves this problem by sending an event hit to Google Analytics (in browsers that support navigator.sendBeacon) when the document is being unloaded. The event category is "Window" and the action is "unload".

Unload does not fire in many cases, see: https://www.igvita.com/2015/11/20/dont-lose-user-and-app-state-use-page-visibility/ - you should listen for visibilitychange instead.

**Comment by philipwalton** [comment]
Thanks @igrigorik, I'll look into this.

**Comment by chaffeqa** [comment]
We've been doing this for all our SAP apps for google analytics.
(mainly based on your [article](https://www.igvita.com/2015/11/20/dont-lose-user-and-app-state-use-page-visibility/) @igrigorik ;)

We wrote a little service for it:

```javascript
var
  VisibilityStates = {
    prerender: 'prerender'
    ,visible: 'visible'
    ,hidden: 'hidden'
  }
  ,VisibilityState = VisibilityStates.visible // dont want to deal with prerender
  ,visibilityCallbacks = {
    visible: []
    ,hidden: []
  }
;
function onForeground(event) {
  processVisibilityChange('visible', event)
}
function onBackground(event) {
  processVisibilityChange('hidden', event)
}
function onVisibilityChange(event) {
  processVisibilityChange(document.visibilityState, event)
}
function processVisibilityChange(newState, event) {
  // console.log("VisibilityState: " + VisibilityState + ' -> ' + VisibilityStates[newState], event)
  // ignore prerender
  if (newState == 'prerender') return;
  // ignore unless changed
  if (VisibilityState == VisibilityStates[newState]) return;
  // fire callbacks
  VisibilityState = VisibilityStates[newState];
  var i, cbs = visibilityCallbacks[newState];
  for (i = 0; i < cbs.length; i++) { cbs[i].call(event) }
}
window.addEventListener("beforeunload", onBackground, false);
window.addEventListener("pagehide", onBackground, false);
window.addEventListener("pageshow", onForeground, false);
if (document.visibilityState) document.addEventListener("visibilitychange", onVisibilityChange, false);
```


So for GA, we would write:

```javascript
visibilityCallbacks.visible.push(function(){
  ga('set', 'page', window.location.href);
  ga('send', 'pageview');
});
```

Obviously you can trim down on the service part, but I'm lazy.

I love that you guys are doing this, finally dont have to write our own custom SAP ga integration on every project 🎉

**Comment by philipwalton** [comment]
Thanks @chaffeqa, there's a `pageVisibility` tracker implementation in the [1.0 branch](https://github.com/googleanalytics/autotrack/tree/1.0-dev). In addition to tracking `visibilitychange` events, it also tracks how long the user spent in each state, and it more gracefully handles session timeouts so you don't end up with a bunch of sessions that only contain `visibilitychange` events.

You can follow the progress of 1.0 in PR #67.

**Comment by igrigorik** [comment]
@chaffeqa nice, great work!

@philipwalton fwiw, it'd be great to have a writeup on the findings in your experiments.. to explain and motivate why folks should be using the visibility approach (within GA and other products). Perhaps something we can collaborate on post I/O?

**Comment by philipwalton** [comment]
Fixed via 911e9f7b7a0d1fe73e00c464c05d57825efbba1b.

@igrigorik agreed, now that 1.0 is out, I'll try to put something together promoting the new model.
